In [5]:
def func(x, y, freq, nop1, nop2):
    N1 = (3+nop1)*(256-x)-1
    N1_ = (3+nop1)*256-1
    N2 = 1 + N1 + (3 + nop2) + (N1_ + 3 + nop2)*(y-1)
    N_a = N2
    print("N1, N1_, N2 =", N1, N1_, N2)
    print(1-25*x+6402*y)
    return N_a/freq


# 54
# 227
# 8
# print('Time for ATMega32: ', 1000*func(213, 241, 8*10**6, 10, 1), 'msec')
print('Time for ATMega328p: ', 1000*func(20, 250, 16*10**6, 22, 0), 'msec')
# print('MAX Time for ATMega328p: ', 1000*func(250, 250, 16*10**6, 30, 30), 'msec')

N1, N1_, N2 = 5899 6399 1600001
1600001
Time for ATMega328p:  100.0000625 msec


In [ ]:
# delay:
# 	LDI R29, 20 ; 
# 	LDI R30, 250 ; 
	
# delay_sub:
# 	INC R29 ;
#     NOP     ; 
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
#     NOP
# 	BRNE delay_sub ;
# 	NOP
# 	DEC R30 ; 
# 	BRNE delay_sub ; 
# 	RET ; 


$$ N_{вн.1} = (1_{INC} + 22_{NOP} + 2_{BRNE})(255-X) + 1_{NOP} + 1_{BRNE} = 25*(256-X)-1 $$
$$ N_{вн.X} = (1_{INC} + 22_{NOP} + 2_{BRNE})(2^{8}-1) + 1_{NOP} + 1_{BRNE} = 25*2^8-1 $$
$$ N_{общ} = 1_{LDI}*2 + (N_{вн.1} + 1_{DEC} + 2_{BRNE}) + (N_{вн.X} + 1_{DEC} + 2_{BRNE})*\\ * (Y-2) + (N_{вн.X} + 1_{DEC} + 1_{BRNE}) = 1 - 25X + 6402Y = 1600002$$
$$ Время\space выполнения: t = N/T = 1600002 / 16*10^6 = 0.1000000625 сек = 100.0000625 мсек$$